From e790200ed5ce420117977336e938535fe08dd246 Mon Sep 17 00:00:00 2001
From: David Benjamin <davidben@google.com>
Date: Thu, 24 Aug 2023 13:04:18 -0400
Subject: [PATCH 11/11] Add BN_bn2lebinpad and BN_lebin2bn

These are OpenSSL names for BN_bn2le_padded and BN_le2bn. We can just
replace BN_le2bn with BN_lebin2bn. BN_bn2lebinpad is not size_t-clean,
so handle it as a separate function like we did BN_bn2binpad.

Change-Id: I6999ca06140a0c8c25942362dc79d1821971d679
Reviewed-on: https://boringssl-review.googlesource.com/c/boringssl/+/62665
Commit-Queue: Adam Langley <agl@google.com>
Auto-Submit: David Benjamin <davidben@google.com>
Reviewed-by: Adam Langley <agl@google.com>
Commit-Queue: David Benjamin <davidben@google.com>
(cherry picked from commit 004317217f81b14c764748f42c5ea88fd05fea3c)

Actually only add new APIs, without changing boundary, and add a
define for the new API name.

Change-Id: I01c26dd32e6a70f6be56a25ff60d650b613e6951
Signed-off-by: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
---
 crypto/bn_extra/convert.c | 8 ++++++++
 include/openssl/bn.h      | 7 +++++++
 2 files changed, 15 insertions(+)

diff --git a/crypto/bn_extra/convert.c b/crypto/bn_extra/convert.c
index 29234effd..c9161fae8 100644
--- a/crypto/bn_extra/convert.c
+++ b/crypto/bn_extra/convert.c
@@ -455,3 +455,11 @@ int BN_bn2binpad(const BIGNUM *in, uint8_t *out, int len) {
   }
   return len;
 }
+
+int BN_bn2lebinpad(const BIGNUM *in, uint8_t *out, int len) {
+  if (len < 0 ||
+      !BN_bn2le_padded(out, (size_t)len, in)) {
+    return -1;
+  }
+  return len;
+}
diff --git a/include/openssl/bn.h b/include/openssl/bn.h
index a03e41f18..47b04079f 100644
--- a/include/openssl/bn.h
+++ b/include/openssl/bn.h
@@ -261,6 +261,7 @@ OPENSSL_EXPORT size_t BN_bn2bin(const BIGNUM *in, uint8_t *out);
 // |BIGNUM| is allocated and returned. It returns NULL on allocation
 // failure.
 OPENSSL_EXPORT BIGNUM *BN_le2bn(const uint8_t *in, size_t len, BIGNUM *ret);
+#define BN_lebin2bn BN_le2bn
 
 // BN_bn2le_padded serialises the absolute value of |in| to |out| as a
 // little-endian integer, which must have |len| of space available, padding
@@ -974,6 +975,12 @@ OPENSSL_EXPORT int BN_MONT_CTX_set(BN_MONT_CTX *mont, const BIGNUM *mod,
 // Use |BN_bn2bin_padded| instead. It is |size_t|-clean.
 OPENSSL_EXPORT int BN_bn2binpad(const BIGNUM *in, uint8_t *out, int len);
 
+// BN_bn2lebinpad behaves like |BN_bn2le_padded|, but it returns |len| on
+// success and -1 on error.
+//
+// Use |BN_bn2le_padded| instead. It is |size_t|-clean.
+OPENSSL_EXPORT int BN_bn2lebinpad(const BIGNUM *in, uint8_t *out, int len);
+
 // BN_prime_checks is a deprecated alias for |BN_prime_checks_for_validation|.
 // Use |BN_prime_checks_for_generation| or |BN_prime_checks_for_validation|
 // instead. (This defaults to the |_for_validation| value in order to be
-- 
2.48.1

